go/cmd/gno_social_feed_indexer/clientql/client.go

```diff
@@ -0,0 +1,120 @@
+package clientql
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/Khan/genqlient/graphql"
+	"github.com/TERITORI/teritori-dapp/go/internal/indexerdb"
+	indexerql "github.com/TERITORI/teritori-dapp/go/pkg/indexerQL"
+	"github.com/TERITORI/teritori-dapp/go/pkg/networks"
+	"github.com/gofrs/uuid"
+	"gorm.io/gorm"
+)
+
+type IndexerQL struct {
+	gqlClient graphql.Client
+	db        *gorm.DB
+	networkID string
+}
+
+func New(networkID string, graphqlEndpoint string, db *gorm.DB) *IndexerQL {
+	gqlClient := graphql.NewClient(graphqlEndpoint, nil)
+	return &IndexerQL{gqlClient: gqlClient, db: db, networkID: networkID}
+}
+
+func (client *IndexerQL) SyncPosts() error {
+	lastPost := client.getLastPost()
+	var fromBlock int
+	if lastPost != nil {
+		fromBlock = int(lastPost.CreatedAt) + 1
+	}
+
+	posts, err := indexerql.GetPostTransactions(context.Background(), client.gqlClient, fromBlock)
+	if err != nil {
+		return err
+	}
+
+	for _, transaction := range posts.Transactions {
+		for _, post := range transaction.Messages {
+			data, ok := post.Value.(*indexerql.GetPostTransactionsTransactionsTransactionMessagesTransactionMessageValueMsgCall)
+			if !ok {
+				fmt.Println("not ok, skipping")
+				continue
+			}
+			post, err := client.getPostWithData(data, int(transaction.Block_height))
+			if err != nil {
+				return err
+			}
+
+			err = client.db.Save(&post).Error
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (client *IndexerQL) getLastPost() *indexerdb.Post {
+	post := &indexerdb.Post{}
+	client.db.Order("created_at desc").Find(post)
+
+	return post
+}
+
+func (client *IndexerQL) getPostWithData(data *indexerql.GetPostTransactionsTransactionsTransactionMessagesTransactionMessageValueMsgCall, blockHeight int) (indexerdb.Post, error) {
+	if len(data.Args) != 4 {
+		return indexerdb.Post{}, fmt.Errorf("invalid args length")
+	}
+	categoryID, _ := strconv.Atoi(data.Args[2])
+	var metadata metadata
+	err := json.Unmarshal([]byte(data.Args[3]), &metadata)
+	if err != nil {
+		return indexerdb.Post{}, err
+	}
+	metadataJSON, err := json.Marshal(metadata)
+	if err != nil {
+		return indexerdb.Post{}, err
+	}
+	post := indexerdb.Post{
+		Identifier:           uuid.Must(uuid.NewV4()).String(),
```

Comment:
we should not use a generated ID but instead use something unique to this post that is consistent after replays

Reply:
Changed it for 
`Identifier:           fmt.Sprintf("%s-%s", data.Caller, transaction.Hash),`

Reply:
this is not unique, what if someone post twice in same tx (currently possible)

we should be able to get the post local id with events or the tx result instead I think

Reply:
Okay I see What do you think of having
`TransactionHash-blockHeight-IndexOfPostInTransaction` ?
maybe just `TransactionHash-IndexOfPostInTransaction` would be enough ?

Reply:
this can work, but it would need to contain network id also since the tx hash could be duplicated on other networks

Reply:
I think the model is not good in the first place, the primary key should be a composite key of `network_id` and `identifier` to correctly be cross-chain

and the `identifier` should only be the on-chain post id

Reply:
this is kinda out of scope, let me think about the best way to resolve this